mobius-extras/src/test/java/com/spotify/mobius/extras/ToggledEventSourceTest.java

```diff
@@ -0,0 +1,158 @@
+  }
+
+  @Test
+  public void testActiveEventSourceUnsubscribesFromBaseWhenLastSubscriberRemoved() {
+    final List<String> received = new ArrayList<>(1);
+
+    togglingEventSource.accept(true);
+    underTest.subscribe(received::add).dispose();
+
+    dataEventSource.assertConsumerCount(0);
+    assertThat(received.size(), equalTo(0));
+  }
+
+  @Test
+  public void testFilterForwardsEventsWhenActive() {
+    final List<String> received = new ArrayList<>(1);
+    togglingEventSource.accept(true);
+    underTest.subscribe(received::add);
+
+    dataEventSource.emit("a");
+
+    dataEventSource.assertConsumerCount(1);
+    assertThat(received.size(), equalTo(1));
+    assertThat(received.get(0), equalTo("a"));
+  }
+
+  @Test
+  public void testToggledEventSourceBlocksEventsAndRemovesConsumersWhenInactive() {
+    final List<String> received = new ArrayList<>(1);
+    togglingEventSource.accept(true);
+    underTest.subscribe(received::add);
+    togglingEventSource.accept(false);
+    dataEventSource.emit("b");
+
+    dataEventSource.assertConsumerCount(0);
+    assertThat(received.size(), equalTo(0));
+  }
+
+  @Test
+  public void testToggledEventSourceDisposeRemovesFromSourceEventSource() {
+    final List<String> received = new ArrayList<>(1);
+    togglingEventSource.accept(true);
+    Disposable d = underTest.subscribe(received::add);
+    d.dispose();
+    dataEventSource.emit("c");
+
+    dataEventSource.assertConsumerCount(0);
+    assertThat(received.size(), equalTo(0));
+  }
+
+  @Test
+  public void testMultipleSubscribersToToggledAllReceiveForwardeEvents() {
+    final List<String> received1 = new ArrayList<>(1);
+    final List<String> received2 = new ArrayList<>(1);
+
+    underTest.subscribe(received1::add);
+    underTest.subscribe(received2::add);
+    togglingEventSource.accept(true);
+
+    dataEventSource.emit("c3");
+
+    assertThat(received1.size(), equalTo(1));
+    assertThat(received1.get(0), equalTo("c3"));
+    assertThat(received2.size(), equalTo(1));
+    assertThat(received2.get(0), equalTo("c3"));
+  }
+
+  @Test
+  public void testMultipleSubscribersAndOneDisposesThenEventsAreStillForwardedToTheOther() {
+    final List<String> received1 = new ArrayList<>(1);
+    final List<String> received2 = new ArrayList<>(1);
+
+    underTest.subscribe(received1::add);
+    Disposable d2 = underTest.subscribe(received2::add);
+    togglingEventSource.accept(true);
+
+    dataEventSource.emit("d41");
+
+    d2.dispose();
+
+    dataEventSource.emit("d42");
+
+    dataEventSource.assertConsumerCount(1); // because our event source is still active
+    assertThat(received1.size(), equalTo(2));
+    assertThat(received1.get(0), equalTo("d41"));
+    assertThat(received1.get(1), equalTo("d42"));
+    assertThat(received2.size(), equalTo(1));
+    assertThat(received2.get(0), equalTo("d41"));
+  }
+
+  @Test
+  public void testThatToggledEventSourceIsNotKeptAliveByUUnderlyingDataSource() {
```

Comment:
Does this test still work if you create a subscription?

Reply:
Yes, it does - just tested. If you subscribe to an EventSource, but don't have a direct reference, it should be free to be GC'd - right?

Reply:
Ok, figured it out, was being kept alive by the Disposable, so once its disposed, it should not be kept alive by a subscription.